tools/src/main/scala/scala/scalanative/linker/Assembly.scala

```diff
@@ -16,14 +14,16 @@ object Assembly {
   final case class Dir private[Assembly](val base: File) extends Assembly {
     //println(s"discovered dir assembly $base")
 
+    def matchFiles(f: File): Array[File] = {
+      if (f.isDirectory) f.listFiles.map(matchFiles).flatten.toArray
+      else if (f.getName.endsWith(".nir")) Array(f)
+      else Array()
+    }
```

Comment:
Linker is actually a very performance sensitive piece code. I used to have dumb non-lazy linker before and it would take minutes to link. I/O is expensive, even with SSDs. If you make a mistake there, it's amplified many times proportionally to number of files / dependencies / etc.